script/makeinstaller.sh

```diff
@@ -262,17 +280,51 @@ StrCmp \$0 \${ffmpeg} "" +2
 [% endblock %]
 EOF
 
-echo "Building Python 3 installer" 1>&2
-
 # copy the streamlinkrc file to the build dir, we cannot use the Include.files property in the config file
 # because those files will always overwrite, and for a config file we do not want to overwrite
-cp "win32/streamlinkrc" "${files_dir}/streamlinkrc"
+cp "${ROOT}/win32/streamlinkrc" "${files_dir}/streamlinkrc"
+
+
+# download binary assets like ffmpeg and rtmpdump from the streamlink assets repo
+# parse the data.json manifest, validate archives and copy specific files to their destination
+log "Fetching assets data from \"${STREAMLINK_ASSETS_REPO}\" (${STREAMLINK_ASSETS_RELEASE})"
+assets_release_data=$(curl -s --fail \
+    -H 'Accept: application/vnd.github.v3+json' \
+    -H "User-Agent: ${GITHUB_REPOSITORY:-"streamlink/streamlink"}" \
+    "https://api.github.com/repos/${STREAMLINK_ASSETS_REPO}/releases/${STREAMLINK_ASSETS_RELEASE}" \
+    || err "Could not fetch release data"
+)
+assets_release_commit=$(echo "${assets_release_data}" | jq -r ".target_commitish")
+assets_data=$(curl -s --fail \
+    -H "User-Agent: ${GITHUB_REPOSITORY:-"streamlink/streamlink"}" \
+    "https://raw.githubusercontent.com/${STREAMLINK_ASSETS_REPO}/${assets_release_commit}/data.json" \
+    || err "Could not fetch manifest data"
+)
+
+log "Retrieving assets"
+while IFS=" " read -r filename size url; do
+    if ! [[ -f "${cache_dir}/${filename}" ]]; then
+        log "Downloading asset: ${filename} (${size} Bytes)"
+        curl -s -L --output "${cache_dir}/${filename}" "${url}"
+    fi
+    checksum=$(jq -r "[.[] | select(.filename == \"${filename}\")] | first | .checksum" <<< "${assets_data}")
+    echo "${checksum} ${cache_dir}/${filename}" | sha256sum --check -
+done < <(jq -r '.assets[] | "\(.name) \(.size) \(.browser_download_url)"' <<< "${assets_release_data}")
+
+log "Extracting assets"
+TEMP=$(mktemp -d) && trap "rm -rf ${TEMP}" EXIT || exit 255
+for ((i=$(jq length <<< "${assets_data}") - 1; i >= 0; --i)); do
+    IFS=" " read -r filename sourcedir targetdir \
+        < <(jq -r ".[$i] | \"\(.filename) \(.sourcedir) \(.targetdir)\"" <<< "${assets_data}")
+    unzip -qq "${cache_dir}/${filename}" -d "${TEMP}"
+    while IFS=" " read -r from to; do
+        log "Copying ${sourcedir}/${from} to ${targetdir}/${to}"
+        install -D -T "${TEMP}/${sourcedir}/${from}" "${files_dir}/${targetdir}/${to}"
+    done < <(jq -r ".[$i].files[] | \"\(.from) \(.to)\"" <<< "${assets_data}")
```

Comment:
I find it a bit nicer to pipe into while, it appears to work the same for me. 

```suggestion
    jq -r ".[$i].files[] | \"\(.from) \(.to)\"" <<< "${assets_data}" | while read -r from to; do
        log "Copying ${sourcedir}/${from} to ${targetdir}/${to}"
        install -D -T "${TEMP}/${sourcedir}/${from}" "${files_dir}/${targetdir}/${to}"
    done
```

Reply:
Piping into while makes it run in a sub-shell, which is what I usually try to avoid.
https://stackoverflow.com/questions/13726764/while-loop-subshell-dilemma-in-bash

Not relevant in this particular case, but I tried to keep the same style throughout the assets and makeinstaller scripts. In the assets script, I ran into an issue because of piping into while.

Reply:
I like it because it isolates the variables, keeps them in their scope. It's easier to read too.

Reply:
Leave it how it is then, it's mostly a style issue for me anyway - I prefer to have the input to at the start of the loop rather than at the end :)
Bash script are hard to read as it is, probably this script would be better off in Python ;D

Reply:
I'll leave it. Otherwise the line gets too long and moving the while statement with the pipe symbol into the next line with added indentation makes it look even more weird.